torchdata/nodes/io/text_streaming_decoder.py

```diff
@@ -0,0 +1,233 @@
+            else:
+                self._current_file = file_data
+                self._source_metadata = {}
+
+            try:
+                # Open the file
+                self._file_handle = open(
+                    self._current_file, self.mode, encoding=self.encoding, transport_params=self.transport_params
+                )
+                self._current_line = 0
+                return True
+            except Exception as e:
+                logger.error(f"Error opening {self._current_file}: {e}")
+                self._file_handle = None
+                return False  # Failed to open file
+
+        except StopIteration:
+            # No more files
+            raise
+
+    def _get_next_line(self) -> Dict:
+        """Read the next line from the current file.
+
+        Returns:
+            Dict: Dictionary with the line data and metadata.
+
+        Raises:
+            StopIteration: If end of file is reached and no more files are available.
+        """
+        try:
+            line = self._file_handle.readline()
+
+            # EOF or empty line at end of file
+            if not line:
+                self._file_handle.close()
+                self._file_handle = None
+                return None  # Signal end of file
+
+            # Create output with metadata
+            metadata = {"file_path": self._current_file, "item_idx": self._current_line}
+
+            # Include metadata from source
+            if self._source_metadata:
+                metadata.update(self._source_metadata)
+
+            self._current_line += 1
+
+            return {self.DATA_KEY: line.rstrip("\n"), self.METADATA_KEY: metadata}
+
+        except Exception as e:
+            logger.error(f"Error reading from {self._current_file}: {e}")
+            if self._file_handle:
+                self._file_handle.close()
+            self._file_handle = None
+            return None  # Signal error
+
+    def next(self) -> Dict:
+        """Get the next line from current file or next available file."""
+        # Loop until we get a valid line or run out of files
+        while True:
+            # If we don't have a file handle, get a new one
+            while self._file_handle is None:
+                if not self._get_next_file():
+                    continue  # Try the next file if this one failed
+
+            # Try to get the next line
+            line_data = self._get_next_line()
+
+            # If we reached the end of the file, try the next one
+            if line_data is None:
+                continue
+
+            # We got a valid line, return it
+            return line_data
+
+    def get_state(self) -> Dict[str, Any]:
+        """Get current state for checkpointing."""
+        return {
+            self.SOURCE_KEY: self.source.state_dict(),
+            self.CURRENT_FILE_KEY: self._current_file,
+            "current_line": self._current_line,
+        }
```

Comment:
shall we also save METADATA_KEY?

Reply:
i) like `self.CURRENT_FILE_KEY`, i added `self.CURRENT_LINE_KEY`. 

ii) are you sure you meant saving `METADATA_KEY`? `DATA_KEY` and `METADATA_KEY` play similar roles which i don't think needs to be saved.